File: InterviewAI/modules/question_generator.py
# Generates HR & Technical questions
import google.generativeai as genai
from typing import List, Dict
from config.settings import GEMINI_API_KEY, GEMINI_MODEL, HR_QUESTIONS_COUNT, TECHNICAL_QUESTIONS_COUNT
from config.prompts import HR_QUESTION_PROMPT, TECHNICAL_QUESTION_PROMPT
from utils.logger import get_logger
from utils.custom_exception import CustomException

logger = get_logger(__name__)

class QuestionGenerator:
    def __init__(self):
        self.logger = logger
        self._setup_gemini()
        
    def _setup_gemini(self):
        """Initialize Gemini API"""
        try:
            if not GEMINI_API_KEY:
                raise CustomException("GEMINI_API_KEY not found. Please set it in your environment variables.")
            
            genai.configure(api_key=GEMINI_API_KEY)
            self.model = genai.GenerativeModel(GEMINI_MODEL)
            self.logger.info("Gemini API initialized successfully")
        except Exception as e:
            self.logger.error(f"Error initializing Gemini API: {str(e)}")
            raise CustomException(f"Failed to initialize Gemini API: {str(e)}")
    
    def generate_hr_questions(self, resume_data: Dict, num_questions: int = None) -> List[Dict]:
        """Generate HR interview questions based on resume"""
        try:
            if num_questions is None:
                num_questions = HR_QUESTIONS_COUNT
            
            self.logger.info(f"Generating {num_questions} HR questions")
            
            # Prepare resume text for prompt
            resume_text = self._prepare_resume_text(resume_data)
            
            # Create prompt
            prompt = HR_QUESTION_PROMPT.format(
                num_questions=num_questions,
                resume_text=resume_text
            )
            
            # Generate questions using Gemini
            response = self.model.generate_content(prompt)
            questions_text = response.text.strip()
            
            # Parse questions
            questions = self._parse_questions(questions_text, "HR")
            
            self.logger.info(f"Successfully generated {len(questions)} HR questions")
            return questions
            
        except Exception as e:
            self.logger.error(f"Error generating HR questions: {str(e)}")
            raise CustomException(f"Failed to generate HR questions: {str(e)}")
    
    def generate_technical_questions(self, resume_data: Dict, num_questions: int = None) -> List[Dict]:
        """Generate technical interview questions based on resume"""
        try:
            if num_questions is None:
                num_questions = TECHNICAL_QUESTIONS_COUNT
            
            self.logger.info(f"Generating {num_questions} technical questions")
            
            # Prepare resume text for prompt
            resume_text = self._prepare_resume_text(resume_data)
            
            # Create prompt
            prompt = TECHNICAL_QUESTION_PROMPT.format(
                num_questions=num_questions,
                resume_text=resume_text
            )
            
            # Generate questions using Gemini
            response = self.model.generate_content(prompt)
            questions_text = response.text.strip()
            
            # Parse questions
            questions = self._parse_questions(questions_text, "Technical")
            
            self.logger.info(f"Successfully generated {len(questions)} technical questions")
            return questions
            
        except Exception as e:
            self.logger.error(f"Error generating technical questions: {str(e)}")
            raise CustomException(f"Failed to generate technical questions: {str(e)}")
    
    def _prepare_resume_text(self, resume_data: Dict) -> str:
        """Prepare resume text for question generation"""
        text_parts = []
        
        # Add contact info
        if resume_data.get('contact_info'):
            contact = resume_data['contact_info']
            if contact.get('email'):
                text_parts.append(f"Email: {contact['email']}")
            if contact.get('phone'):
                text_parts.append(f"Phone: {contact['phone']}")
        
        # Add sections
        if resume_data.get('sections'):
            for section_name, content in resume_data['sections'].items():
                text_parts.append(f"{section_name.title()}: {content}")
        
        # Add skills
        if resume_data.get('skills'):
            skills_text = ", ".join(resume_data['skills'])
            text_parts.append(f"Skills: {skills_text}")
        
        # Add raw text if sections are not available
        if not text_parts and resume_data.get('raw_text'):
            text_parts.append(resume_data['raw_text'])
        
        return "\n\n".join(text_parts)
    
    def _parse_questions(self, questions_text: str, question_type: str) -> List[Dict]:
        """Parse generated questions text into structured format"""
        questions = []
        lines = questions_text.split('\n')
        
        for line in lines:
            line = line.strip()
            if not line:
                continue
            
            # Remove numbering (1., 2., etc.)
            import re
            line = re.sub(r'^\d+\.\s*', '', line)
            
            if line and len(line) > 10:  # Basic validation
                questions.append({
                    'question': line,
                    'type': question_type,
                    'difficulty': self._assess_difficulty(line),
                    'category': self._categorize_question(line, question_type)
                })
        
        return questions
    
    def _assess_difficulty(self, question: str) -> str:
        """Assess question difficulty based on content"""
        question_lower = question.lower()
        
        # Simple heuristic for difficulty assessment
        if any(word in question_lower for word in ['explain', 'describe', 'analyze', 'compare', 'evaluate']):
            return 'Hard'
        elif any(word in question_lower for word in ['what', 'how', 'why', 'when', 'where']):
            return 'Medium'
        else:
            return 'Easy'
    
    def _categorize_question(self, question: str, question_type: str) -> str:
        """Categorize question based on content"""
        question_lower = question.lower()
        
        if question_type == "HR":
            if any(word in question_lower for word in ['team', 'teamwork', 'collaboration']):
                return 'Teamwork'
            elif any(word in question_lower for word in ['lead', 'leadership', 'manage']):
                return 'Leadership'
            elif any(word in question_lower for word in ['problem', 'challenge', 'difficult']):
                return 'Problem Solving'
            elif any(word in question_lower for word in ['goal', 'career', 'future']):
                return 'Career Goals'
            else:
                return 'General HR'
        
        else:  # Technical
            if any(word in question_lower for word in ['code', 'programming', 'algorithm']):
                return 'Programming'
            elif any(word in question_lower for word in ['database', 'sql', 'data']):
                return 'Database'
            elif any(word in question_lower for word in ['system', 'architecture', 'design']):
                return 'System Design'
            elif any(word in question_lower for word in ['framework', 'library', 'tool']):
                return 'Frameworks & Tools'
            else:
                return 'General Technical'

File: server/services/questionGenerator.js
const OpenAI = require('openai');

class QuestionGenerator {
  constructor() {
    this.openai = new OpenAI({
      apiKey: process.env.OPENAI_API_KEY
    });
  }

  async generateQuestions(resumeData, type, count = 5) {
    try {
      const prompt = this.createPrompt(resumeData, type, count);
      
      const response = await this.openai.chat.completions.create({
        model: "gpt-3.5-turbo",
        messages: [
          {
            role: "system",
            content: "You are an expert interview coach who generates personalized interview questions based on resume data. Generate questions that are relevant, challenging, and help assess the candidate's skills and experience."
          },
          {
            role: "user",
            content: prompt
          }
        ],
        temperature: 0.7,
        max_tokens: 2000
      });

      const questionsText = response.choices[0].message.content;
      return this.parseQuestions(questionsText, type);
    } catch (error) {
      console.error('OpenAI API error:', error);
      throw new Error(`Failed to generate questions: ${error.message}`);
    }
  }

  createPrompt(resumeData, type, count) {
    const resumeText = this.prepareResumeText(resumeData);
    
    if (type === 'HR') {
      return `Based on the following resume information, generate ${count} HR interview questions that would be appropriate for this candidate.

Resume Information:
${resumeText}

The questions should cover:
1. Behavioral questions about past experiences
2. Motivation and career goals
3. Teamwork and leadership
4. Problem-solving abilities
5. Cultural fit and values

Generate questions that are:
- Relevant to the candidate's background
- Open-ended to encourage detailed responses
- Professional and appropriate
- Varied in difficulty and topic

Return the questions in JSON format with the following structure:
[
  {
    "id": "unique_id",
    "question": "Question text here",
    "type": "HR",
    "difficulty": "Easy|Medium|Hard",
    "category": "Category name"
  }
]`;
    } else {
      return `Based on the following resume information, generate ${count} technical interview questions that would be appropriate for this candidate.

Resume Information:
${resumeText}

The questions should cover:
1. Technical skills mentioned in the resume
2. Programming languages and frameworks
3. Problem-solving and coding challenges
4. System design concepts (if applicable)
5. Industry-specific knowledge

Generate questions that are:
- Relevant to the candidate's technical background
- Appropriate for their experience level
- Mix of conceptual and practical questions
- Varied in difficulty

Return the questions in JSON format with the following structure:
[
  {
    "id": "unique_id",
    "question": "Question text here",
    "type": "Technical",
    "difficulty": "Easy|Medium|Hard",
    "category": "Category name"
  }
]`;
    }
  }

  prepareResumeText(resumeData) {
    const textParts = [];

    // Add contact info
    if (resumeData.contactInfo) {
      const contact = resumeData.contactInfo;
      if (contact.name) textParts.push(`Name: ${contact.name}`);
      if (contact.email) textParts.push(`Email: ${contact.email}`);
      if (contact.phone) textParts.push(`Phone: ${contact.phone}`);
    }

    // Add sections
    if (resumeData.sections) {
      for (const [sectionName, content] of Object.entries(resumeData.sections)) {
        textParts.push(`${sectionName.charAt(0).toUpperCase() + sectionName.slice(1)}: ${content}`);
      }
    }

    // Add skills
    if (resumeData.skills && resumeData.skills.length > 0) {
      const skillsText = resumeData.skills.join(', ');
      textParts.push(`Skills: ${skillsText}`);
    }

    // Add raw text if sections are not available
    if (textParts.length === 0 && resumeData.rawText) {
      textParts.push(resumeData.rawText);
    }

    return textParts.join('\n\n');
  }

  parseQuestions(questionsText, type) {
    try {
      // Try to parse as JSON first
      const questions = JSON.parse(questionsText);
      
      // Validate and format questions
      return questions.map((q, index) => ({
        id: q.id || `q_${type.toLowerCase()}_${index + 1}`,
        question: q.question || '',
        type: type,
        difficulty: q.difficulty || this.assessDifficulty(q.question),
        category: q.category || this.categorizeQuestion(q.question, type)
      }));
    } catch (error) {
      // Fallback: parse from text format
      return this.parseQuestionsFromText(questionsText, type);
    }
  }

  parseQuestionsFromText(questionsText, type) {
    const questions = [];
    const lines = questionsText.split('\n').filter(line => line.trim().length > 0);
    
    let questionIndex = 1;
    for (const line of lines) {
      // Remove numbering (1., 2., etc.)
      const cleanLine = line.replace(/^\d+\.\s*/, '').trim();
      
      if (cleanLine.length > 10 && cleanLine.includes('?')) {
        questions.push({
          id: `q_${type.toLowerCase()}_${questionIndex}`,
          question: cleanLine,
          type: type,
          difficulty: this.assessDifficulty(cleanLine),
          category: this.categorizeQuestion(cleanLine, type)
        });
        questionIndex++;
      }
    }

    return questions;
  }

  assessDifficulty(question) {
    const questionLower = question.toLowerCase();
    
    if (questionLower.includes('explain') || 
        questionLower.includes('describe') || 
        questionLower.includes('analyze') || 
        questionLower.includes('compare') || 
        questionLower.includes('evaluate')) {
      return 'Hard';
    } else if (questionLower.includes('what') || 
               questionLower.includes('how') || 
               questionLower.includes('why') || 
               questionLower.includes('when') || 
               questionLower.includes('where')) {
      return 'Medium';
    } else {
      return 'Easy';
    }
  }

  categorizeQuestion(question, type) {
    const questionLower = question.toLowerCase();
    
    if (type === 'HR') {
      if (questionLower.includes('team') || questionLower.includes('teamwork') || questionLower.includes('collaboration')) {
        return 'Teamwork';
      } else if (questionLower.includes('lead') || questionLower.includes('leadership') || questionLower.includes('manage')) {
        return 'Leadership';
      } else if (questionLower.includes('problem') || questionLower.includes('challenge') || questionLower.includes('difficult')) {
        return 'Problem Solving';
      } else if (questionLower.includes('goal') || questionLower.includes('career') || questionLower.includes('future')) {
        return 'Career Goals';
      } else {
        return 'General HR';
      }
    } else {
      if (questionLower.includes('code') || questionLower.includes('programming') || questionLower.includes('algorithm')) {
        return 'Programming';
      } else if (questionLower.includes('database') || questionLower.includes('sql') || questionLower.includes('data')) {
        return 'Database';
      } else if (questionLower.includes('system') || questionLower.includes('architecture') || questionLower.includes('design')) {
        return 'System Design';
      } else if (questionLower.includes('framework') || questionLower.includes('library') || questionLower.includes('tool')) {
        return 'Frameworks & Tools';
      } else {
        return 'General Technical';
      }
    }
  }
}

module.exports = new QuestionGenerator();
